### DISCUSSION TITLE
Creating Tiers of Projects within the MapLibre Organization

### DISCUSSION THREAD

**Discussion by lseelenbinder** [maplibre/maplibre#132] [Ideas]
Hi all! 👋

I'd like to suggest a small formalization to our current system of different projects to encourage more participation, make our goals, priorities, and commitments as a governing board, organization, and broader community clear.

### The Problem

Currently, we have a lot of projects (26 repositories at last count) under the @maplibre organization. These include everything from the keystone GL JS & Native projects to plugins, libraries, and SDKs built around the GL projects. However, it is not clear from the outside perspective what projects we formally support (dedicate resources to, are committed to ensuring their long-term success, etc.) and what projects are _hosted_ for the community to support as they have time and need.

This problem is further complicated by the fact that when someone wants to onboard a new project, the board is hesitant to do that if we don't see a clear way to support it with the resources we have. This creates a bit of a catch-22 where we may not have an "official" SDK for a platform where we should, but we can't get community unity on a SDK because there is not single version under the MapLibre project, and contributes to the fragmentation the MapLibre organization was created to avoid.

### Proposed Solution

I propose we introduce **Project Tiers**, with the following initial tiers:

- **Core Project** - this is a project that the organization is committed to in a long-term way, and will provide resources to develop, potentially maintainers to work on, etc. Examples: MapLibre GL JS & MapLibre GL Native
- **Supported Project** - this is a project the organization is supporting, but cannot guarantee any level of long-term support or resources. Examples: `maplibre-rs`
- **Hosted Project** - this is a project for which the community is responsible for all maintenance and support; the organization may choose to "promote" it at some future time, but make no guarantees. Examples: `martin`, some future React Native SDK

I think this creates clear expectations across all of our projects, while allowing us to provide for the success of the whole MapLibre ecosystem.

I'm eager to hear your feedback!

**Comment by HarelM** [comment]
I think these definitions are great in order to understand what to expect from a project I'm using (as a developer).
I'm curious as to how a project can be "selected" to be in a specific tier? What's the right process for that? If a project was selected to be in a certain tier what is the process to request for a reevaluation?
Can the TSC influence the decision or is it only up to the board to decide? Can/should this be voted on?

**Reply by wipfli** [reply]
Maybe becoming a core project could require approval from the MapLibre Voting Members...

**Reply by lseelenbinder** [reply]
Good questions, @HarelM.

Here's my suggested approach, intended to reduce bureaucracy to a minimum:
- Projects start on the "lowest" tier and can then be nominated by voting members for consideration by the board via an issue/slack message/etc. The board will vote on it at the next meeting. If voting members wish to override the board's decision, they may do so with a 2/3rds majority of a quorum.
- Reevaluation happens with the same process, within reason.

Becoming a core project is a bit more weighty, so I like @wipfli's idea of requiring the voting members to "adopt" it.

> Can the TSC influence the decision or is it only up to the board to decide? Can/should this be voted on?

This is an interesting question. I would say, first, the standing TSCs will not dedicate meeting time on the non-supported/core projects; or at the very least, at the expense of talking about supported/core projects. The community-led projects may form their own TSC, but won't be official.

If the TSC sees a project as vital, and deserving of moving to a supported or core position, they may choose to nominate it, just like any other voting member.

**Comment by wipfli** [comment]
I like it, @lseelenbinder. I think we could also just not give a label to projects which are not receiving donated resources, because I think all of our projects should be community-led and we just use donated resources to help the community achieve their goals.

**Reply by lseelenbinder** [reply]
Good point, @wipfli. My intent was not to say the other projects _aren't_ community-led (because they definitely are!), but to indicate that the other projects are _only_ community-led. There is no official support or resources from donations, etc. being given to them.

I'm open to another term, for sure, but I think we should label them all with something. That gives us a consistent message.

**Reply by wipfli** [reply]
What about "hosted" instead of "community-led"?

**Reply by lseelenbinder** [reply]
I like it!

**Reply by wipfli** [reply]
Because the MapLibre umbrella hosts the project...

**Comment by maxammann** [comment]
Great idea!

## Status quo

When I updated the website I kind of created such a tier-system implicitly: https://maplibre.org/projects/

In that case there are three kinds of projects:
* Unstable projects (denoted by the badge on the website)
* Stable projects
* Not-listed projects

## My 2 cents

In order to simplify this tier system I would propose another classification though. I think what matters to users is if a project is supported or not supported. Having the classification of core vs non-core does probably not help users and it will cause discussions :D

## Alternative 2D system 

I would like to propose to have a 2 dimensional system. While that sounds more complex, it maybe depict better what users are looking for. The 2 dimensions are: Supported vs. Not-supported, Stable vs. Unstable


E.g:

* martin
  |             | Stable | Unstable |
  |-------------|--------|----------|
  | Supported   |        | x        |
  | Unsupported |        |          |

All 4 possibilities are possible.

**Reply by HarelM** [reply]
What do you define as an unstable project? Why would we have those laying around?

**Reply by maxammann** [reply]
Unstable projects usually:
* don't have a good test coverage,
* miss production ready releases, and
* are not packaged for different operating systems

That is true for maplibre-rs. martin is currently at the verge of becoming stable in my opinion.

As long as we allow new projects to be started under maplibre, we will have unstable projects.

**Reply by HarelM** [reply]
I would argue that this should be a criteria to be considered in order to be in maplibre, i.e. add these requirements to the checklist. I don't think we should have unstable project under the umbrella as it might affect the overall perception of maplibre.

**Reply by wipfli** [reply]
Thanks for this perspective, @maxammann. I think knowing what to expect from a project is useful for people who are approaching it.

The tier system aims to solve a different problem though, which to set expectations when it comes to allocation of donated resources. We have to be clear on which projects we can use Maintainer time and on which projects we can use Bounties. I.e. I would like to clarify how the Developer and Maintainer roles map to projects in the MapLibre Organzation.

**Reply by lseelenbinder** [reply]
Thanks for your input @maxammann! I think we could have a 2D system, because, as you say, it's already somewhat status quo.

I think we should prefer an "experimental" label for projects that aren't ready for primetime, but related to the long-term interests of the MapLibre community. This is a subtle difference from a stable/unstable label, but indicates the goals better, I think. An experimental project could be on the path towards "production" ready, but is not there yet, or may never get there and will be archived because the experiment proved unhelpful.

That being said, let's keep this discussion focused on tiers, because I think we can discuss these items separately (they are orthogonal), and mixing the conversation will make the decision a lot harder. 😄

**Reply by wipfli** [reply]
+1 for experimental

**Reply by maxammann** [reply]
experimental works for me too 😃

**Reply by maxammann** [reply]
> I would argue that this should be a criteria to be considered in order to be in maplibre, i.e. add these requirements to the checklist. I don't think we should have unstable project under the umbrella as it might affect the overall perception of maplibre.

I don't think so. If experimental projects are not allowed, then no new projectscan be started.

**Reply by HarelM** [reply]
I'm stopping the discussion about experimental projects as it's not related to this thread. I'll write it down to make sure we talk about it in either the TSC or the governing board.

**Reply by maxammann** [reply]
Yeah sorry, I did not get that the initial problem was related to the allocation of resources, instead of defining the status of the different projects for maplibre users.

Those are definitely two separate topics!

**Comment by wipfli** [comment]
I am preparing at the moment the budget for the months Dec 2022, Jan 2023, Feb 2023 which will have to be approved by the MapLibre Governing Board. The tiers system you propose here @lseelenbinder would be very helpful for the budget I think. We could do the following:

* **Core Project**: Gets Maintainer time and Bounties
* **Supported Project**: Gets Bounties
* **Hosted Project**: Gets no donated resources

Does this align with your perspective?

**Reply by lseelenbinder** [reply]
Yes, but I would write it as:

- Core Project: Receives Maintainer time and Bounties
- Supported Project: May be allocated money for Bounties
- Hosted Project: Gets no donated resources

**Comment by roblabs** [comment]
A good example of _project_  to be considered for these proposed Tiers would be Style Editors.  

Style Editors are tooling and an important part of the map pipeline.  For example "_the Style generated by the Style Editing Tool should work & look well with MapLibre JS or Native rendering._  Style Editors today do not use MapLibre for rendering.  Imagine the case of complex expressions that may be in the Style Editor, but not supported in the rendering SDK.

Maybe a Style Editor is not a _core_ project, maybe it is in the _supported_ tier, but it should certainly be considered in the _hosted_ tier.

---

For more information on Style Editors and related discussions on upgrading to MapLibre:
* https://github.com/maputnik/editor/issues/788
* https://github.com/go-spatial/fresco/issues/133

**Reply by maxammann** [reply]
I absolutely agree! Through I belive there is currently not much drive behind the style editors. Maybe those things will also just become a thing if maplibre-gl-js diverges more from mapbox and issues become evident.
Right now I believe that it just works well enough: Using mapbox in the style editor and maplibre in production.

**Comment by wipfli** [comment]
I think this is a great input from your side @lseelenbinder and I would like to conclude the discussion by proposing a process for projects tiers. I will write a wiki page about it I think, or is it better if we have a `PROJECT_TIERS.md` document in `maplibre/maplibre`?

**Reply by wipfli** [reply]
I went for the markdown at #152...
